### PR TITLE
Added proxy support for XEP-0363

### PR DESCRIPTION
**Pull request by Vshnv**
HttpFileUploadManager did not use proxy info from connection to open the connection. Changed it so it uses the proxy info from the XMPP connection config to connect through the specified proxy where possible.

### CODE REVIEW

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 1]
```diff
@@ -502,6 +507,46 @@ private void upload(InputStream iStream, long fileSize, Slot slot, UploadProgres
         }
     }
 
+    private static HttpURLConnection createURLConnection(XMPPConnection connection, URL putUrl) throws IOException {
+        Objects.requireNonNull(putUrl);
+        ProxyInfo proxyInfo = fetchProxyInfo(connection);
+        if (proxyInfo != null) {
+            return createProxiedURLConnection(proxyInfo, putUrl);
+        } else {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:515]: Drop the `else` here, as you return in the `if` case.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java [hunk 2]
```diff
@@ -502,6 +507,46 @@ private void upload(InputStream iStream, long fileSize, Slot slot, UploadProgres
...
+
+    private static Proxy.Type convertProxyType(ProxyInfo.ProxyType type) {
+        switch (type) {
+            case HTTP:
+                return Proxy.Type.HTTP;
+            case SOCKS4:
+            case SOCKS5:
+                return Proxy.Type.SOCKS;
+        }
+        return Proxy.Type.DIRECT;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:539]: This can never happen, right? So I'd add a `default` case and throw an `AssertionError` in this case. I also think that this method should be added to `ProxyInfo`, e.g. as `ProxyInfo.getJavaNetProxyType()`.

**Review by Flowdalic — CHANGES_REQUESTED** [review]
Thanks for your contribution. This appears to be a sensible addition. Please find my comments inline.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java (hunk 2 above)
Comment (Vshnv) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java:539]: Makes sense. Ill make the required changes and push in a bit.

File: smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java [hunk 3]
```diff
@@ -24,17 +28,18 @@
  */
 
 public class ProxyInfo {
+
```
Comment (Flowdalic) [inline smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java:31]: Please do not include unrelated whitespace changes in a commit. If you really think there should be a newline, then create an extra commit and pull request for this.

File: smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java [hunk 4]
```diff
@@ -24,17 +28,18 @@
...
         SOCKS4,
         SOCKS5
     }
 
-    private String proxyAddress;
-    private int proxyPort;
-    private String proxyUsername;
-    private String proxyPassword;
-    private ProxyType proxyType;
+    private final String proxyAddress;
```
Comment (Flowdalic) [inline smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java:38]: This and the following changes adding `final` seem very sensible, but should be an extra commit and pull request.

File: smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxySocketConnection.java [hunk 5]
```diff
@@ -34,4 +36,17 @@
...
+    static Function<ProxySocketConnection, ProxyInfo> forProxyType(ProxyInfo.ProxyType proxyType) {
+        switch (proxyType) {
+            case HTTP:
+                return HTTPProxySocketConnection::new;
+            case SOCKS4:
+                return Socks4ProxySocketConnection::new;
+            case SOCKS5:
+                return Socks5ProxySocketConnection::new;
+            default:
+                throw new IllegalStateException();
```
Comment (Flowdalic) [inline smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxySocketConnection.java:49]: If anything this should be an `AssertionError("Unknown proxy type: " + proxyType)`. But since you cover all cases of the`ProxyInfo.ProxyType` enum, do you need a default case at all?

File: smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java [hunk 6]
```diff
@@ -74,6 +70,18 @@ public static ProxyInfo forSocks5Proxy(String pHost, int pPort, String pUser,
...
 
+    public Proxy.Type getJavaProxyType() {
+        switch (proxyType) {
+            case HTTP:
+                return Proxy.Type.HTTP;
+            case SOCKS4:
+            case SOCKS5:
+                return Proxy.Type.SOCKS;
+            default:
+                throw new AssertionError("Unreachable case.");
```
Comment (Flowdalic) [inline smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java:81]: Exception messages should always strive to include sensible context, e.g. in this case `"Unkown proxy type: "  + proxyType`. But since you cover all cases of theProxyInfo.ProxyType enum, do you need a default case at all?

**Review by Flowdalic — CHANGES_REQUESTED** [review]
Thanks, this is clearly an improvement over the first version.

But please see if you can not rebase the PR into two commits: One that makes the `o.j.smack.proxy` related changes, and one that makes the `o.j.smackx.httpfileupload` related changes.

If I am not mistaken, then this would cleanly apply to the 4.4 branch, and also be a good candidate for the 4.4. branch. So maybe consider rebasing your commits onto the 4.4 branch and changing the target branch of this PR to 4.4. It should be trivial.

File: smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java (hunk 3 above)
Comment (Vshnv) [reply smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java:31]: Ah, my bad. Did not notice the extra line. Reverting that

File: smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java (hunk 4 above)
Comment (Vshnv) [reply smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxyInfo.java:38]: Alrighty, will do

File: smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxySocketConnection.java (hunk 5 above)
Comment (Vshnv) [reply smack-core/src/main/java/org/jivesoftware/smack/proxy/ProxySocketConnection.java:49]: Ah, I just move the IllegalStateException part from the other constructor as is, Ill make the change to an Assertion error. 

> do you need a default case at all?

Its just considered easier to understand exhaustive cases without going into the enum declaration with the explicit exception.

**Comment by Flowdalic** [comment]
It appears your rebate went wrong, there should only be the two commits of you in the PR. Hint: maybe hard reset the source branch of this PR to 4.4, then cherry pick the two commits on top of it and then force push.

**Comment by Vshnv** [comment]
Ah alright, thanks, was a bit stuck with that.

**Comment by Vshnv** [comment]
Completed requested changes. Re-opening due to auto-close from the force-push.

**Comment by Flowdalic** [comment]
Much better. Muchas gracias :)
